src/converters.ts

```diff
@@ -184,25 +184,65 @@ const boolean = new Converter<boolean, boolean>({
 });
 
 export interface DecimalOptions {
-  maxWholeDigits?: number;
-  decimalPlaces?: number;
-  allowNegative?: boolean;
+  maxWholeDigits: number;
+  decimalPlaces: number;
+  allowNegative: boolean;
 }
 
-function decimal(options?: DecimalOptions) {
-  const maxWholeDigits: number =
-    options == null || !options.maxWholeDigits ? 10 : options.maxWholeDigits;
-  const decimalPlaces: number =
-    options == null || !options.decimalPlaces ? 2 : options.decimalPlaces;
-  const allowNegative: boolean =
-    options == null || options.allowNegative == null
-      ? true
-      : options.allowNegative;
+function decimal(
+  options?:
+    | Partial<DecimalOptions>
+    | ((context: any) => Partial<DecimalOptions>)
+) {
+  function getOptions(context: any): DecimalOptions {
+    if (typeof options === "function") {
+      return getDecimalOptions(options(context));
+    }
+    if (options == null) {
+      return { maxWholeDigits: 10, decimalPlaces: 2, allowNegative: true };
+    }
+    return getDecimalOptions(options);
+  }
 
-  const regex = new RegExp(
-    `^${allowNegative ? `-?` : ``}(0|[1-9]\\d{0,${maxWholeDigits -
-      1}})(\\.\\d{0,${decimalPlaces}})?$`
-  );
+  function getDecimalOptions(options: Partial<DecimalOptions>) {
+    const maxWholeDigits: number = options.maxWholeDigits || 10;
+    const decimalPlaces: number =
+      options.decimalPlaces == null ? 2 : options.decimalPlaces;
+    const allowNegative: boolean =
+      options.allowNegative == null ? true : options.allowNegative;
+    return { maxWholeDigits, decimalPlaces, allowNegative };
+  }
+
+  function getRegex(context: any): RegExp {
+    const options = getOptions(context);
+    return new RegExp(
+      `^${
+        options.allowNegative ? `-?` : ``
+      }(0|[1-9]\\d{0,${options.maxWholeDigits - 1}})(\\.\\d{0,${
+        options.decimalPlaces
+      }})?$`
+    );
+  }
+
+  function trimDecimals(
```

Comment:
There are a few ugly things in this function. That you re-assign to splitValue[1] for instance. And you refer to it multiple times. I prefer to be very explicit and not modify existing data if I can avoid it.

I suggest something like following:

```js
const [before, after] = value.split(".")
if (after === "undefined") {
   return value;
}
const trimmedAfter = after.substring(....);
if (trimmedAfter.length === 0) {
    return before;
} 
return [before, trimmedAfter].join('.')
```